### PR TITLE
server: support gzip compression for metrics (#12356)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #12356 to release-6.1
You can switch your code base to this Pull Request by using [git-extras](https://github.com/tj/git-extras):
```bash
# In tikv repo:
git pr https://github.com/tikv/tikv/pull/12647
```

After apply modifications, you can push your change to this PR via:
```bash
git push git@github.com:ti-srebot/tikv.git pr/12647:release-6.1-47af4d62d8b0
```

---

Signed-off-by: glorv <glorvs@163.com>

### What is changed and how it works?

Issue Number: ref #12355

What's Changed:
Support write metrics response with gzip compression. If the request header contains `Accept-Encoding: gzip`, tikv server will return the response with gzip compressed; If not, still return no-compressed response as before.

In my test in a local environment, use gzip compression can reduce the body size to about 1/12.
```
$ curl -v -i --compressed -s 127.0.0.1:20180/metrics
...
< HTTP/1.1 200 OK
< content-encoding: gzip
< content-type: text/plain; charset=utf-8
< content-length: 118739
< date: Mon, 18 Apr 2022 08:59:22 GMT

$ curl -v -i -s 127.0.0.1:20180/metrics
< HTTP/1.1 200 OK
< content-type: text/plain; charset=utf-8
< content-length: 1430175
< date: Mon, 18 Apr 2022 08:58:24 GMT
```

Performance Impact:
The gzip compression will cost about 12ms extra cpu time, which should not be the bottleneck in a typical 8c environment.

In my benchmark with (1*tidb(40c), 1*pd, 3*tikv(8c,16gb)) with 3k tpcc workload. With a concurrency of 200, each tikv's cpu usage is 700+, compare between v6.0.0 and this commit(based on v6.0.0), there is no visiable performance impact for all of tmpC, P99, P999 metrics.  

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

### Release note 

```release-note
Support encode metrics response with gzip to reduce the http body size
```

### CODE REVIEW

**Review by glorv — APPROVED** [review]

File: components/tikv_util/src/metrics/mod.rs [hunk 1]
```diff
@@ -31,21 +34,22 @@ pub use self::metrics_reader::HistogramReader;
...
 
-use std::collections::HashMap;
-
-use kvproto::pdpb;
 pub type RecordPairVec = Vec<pdpb::RecordPair>;
 
 pub fn dump() -> String {
     let mut buffer = vec![];
+    dump_to(&mut buffer);
+    String::from_utf8(buffer).unwrap()
```
Comment (zhangjinpeng1987) [inline components/tikv_util/src/metrics/mod.rs:42]: Is there twice mem copy?

Comment (glorv) [reply components/tikv_util/src/metrics/mod.rs:42]: No. the buffer is moved into the string.

File: src/server/status_server/mod.rs [hunk 2]
```diff
@@ -495,6 +502,27 @@ where
...
+        let metrics = if gz_encoding {
+            // gzip can reduce the body size to less than 1/10.
+            let mut encoder = GzEncoder::new(vec![], Compression::default());
+            dump_to(&mut encoder);
+            encoder.finish().unwrap()
+        } else {
+            dump().into_bytes()
+        };
+        let mut resp = Response::new(metrics.into());
+        resp.headers_mut()
```
Comment (zhangjinpeng1987) [inline src/server/status_server/mod.rs:516]: Why not combine this branch with the `if` below?

**Review by zhangjinpeng1987 — COMMENTED** [review]
rest LGTM

File: src/server/status_server/mod.rs (hunk 2 above)
Comment (glorv) [reply src/server/status_server/mod.rs:516]: the `CONTENT-TYPE` header is always set, `CONTENT-ENCODING` is set only when it is gzip.

**Comment by glorv** [comment]
@BusyJay  PTAL, thank~

**Review by BusyJay — APPROVED** [review]

**Pull request merged**